File: sections/23_knapsack_subset_dp/problems/c_possible_sums/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    a = data[1:1 + n]
    total = sum(a)
    possible = [False] * (total + 1)
    possible[0] = True
    current = 0
    for x in a:
        for s in range(current, -1, -1):
            if possible[s]:
                possible[s + x] = True
        current += x
    ans = [i for i in range(1, total + 1) if possible[i]]
    print(len(ans))
    print(" ".join(map(str, ans)))


if __name__ == "__main__":
    main()

File: sections/23_knapsack_subset_dp/problems/c_possible_sums/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<int> a(n);
    int total = 0;
    for (int &x : a) {
        cin >> x;
        total += x;
    }
    vector<char> possible(total + 1, false);
    possible[0] = true;
    int current = 0;
    for (int x : a) {
        for (int s = current; s >= 0; s--) {
            if (possible[s]) possible[s + x] = true;
        }
        current += x;
    }
    vector<int> ans;
    for (int s = 1; s <= total; s++) {
        if (possible[s]) ans.push_back(s);
    }
    cout << ans.size() << '\n';
    for (int i = 0; i < (int)ans.size(); i++) {
        if (i) cout << ' ';
        cout << ans[i];
    }
    cout << '\n';
    return 0;
}

